[1.6.x] Android native copy files

Laserlicht opened the pull request:
attempt to fix #5193

IvanSavenko commented:
Not required, but more details logging / error reporting would also be nice (does copied file exists? Does it have correct size? Can we open it? ) - so in case if something still does not works, we would have some useful info about what exactly is broken.

kambala-decapitator commented on launcher/firstLaunch/firstlaunch_moc.cpp:
```diff
@@ -369,8 +371,20 @@ void FirstLaunchView::extractGogData()
+#ifdef VCMI_ANDROID
+		auto copyFunc = [](std::string src, std::string dst)
+		{
+			CAndroidVMHelper vmHelper;
+			vmHelper.callCustomMethod(CAndroidVMHelper::NATIVE_METHODS_DEFAULT_CLASS, "copyFileFromUri", "(Ljava/lang/String;Ljava/lang/String;)V", [src, dst](JNIEnv * env, jclass javaHelper, jmethodID methodId){
+				env->CallStaticVoidMethod(javaHelper, methodId, env->NewStringUTF(src.c_str()), env->NewStringUTF(dst.c_str()));
+			}, true);
```
not sure why you decided to write something out of your head instead of checking how it's already done in the launcher, for example:
https://github.com/vcmi/vcmi/blob/6e032946337ccaaeb189cc5f01d64f8299068305/launcher/prepare.cpp#L63

and it doesn't involve any lambdas, just pass proper objects

Laserlicht replied:
To be honest. I wasn't aware of this function. I thought all functions that are called by native c++ code would be in NativeFunctions.

Feel free to provide tested code to use suggested approach.

I don't have any android build environment on my pc. The build instructions for android never worked for me. And without working ccache it's a pain to use CI for this.
Don't want to spend another hours on this. Was already too time consuming. But we needed a solution for this.

kambala-decapitator replied:
> I don't have any android build environment on my pc. The build instructions for android never worked for me

have you tried asking for help / guidance?

> And without working ccache it's a pain to use CI for this.

are you in a hurry? We're not releasing another update in a few hours :) I suggest to configure local environment to speed things up.

Laserlicht replied:
> are you in a hurry?

Not really.
But prioritys are shifting. Winter holiday season is over. So not sure how much time left in the next weeks besides small fixes. And would be great to fix this in the next 1.6.x as we have many reports on this problem.

kambala-decapitator commented on android/vcmi-app/src/main/java/eu/vcmi/vcmi/NativeMethods.java:
```diff
@@ -104,4 +111,29 @@ private static Messenger requireServerMessenger()
     }
+
+    @SuppressWarnings(Const.JNI_METHOD_SUPPRESS)
+    public static void copyFileFromUri(String sourceFileUri, String destinationFile)
```
it should go into `util/FileUtil.java`. And that file already contains basically the same code inside the directory copy method, just extract it.

kambala-decapitator commented:
ignoring review comments is not a healthy way of collaboration. I'm against merging PR in the current state.

Laserlicht commented:
> ignoring review comments is not a healthy way of collaboration. I'm against merging PR in the current state.

Insinuations are also not helpful. I was not ignoring your comments.

But my usual way is solving problems before open 10 new problems.

I wanted to know if it in general works. And for that a prototype build is needed. Because if it didn't work it's not worth to spend any time on it.

And the android build isn't that easy like e.g. for Linux. For non android devs the docs arent also not really clear (how to install sdk etc.). For example compiling fuzzylite doesn't work. Seen your comments in forum. But don't want mess with my system cmake install. Now my way is a docker container... Btw. why do we not provide dockerfiles for building? Would be much easier with this complex building mechanism.

kambala-decapitator commented:
> I was not ignoring your comments (that's why I left them open).

well, you responded with "I didn't know about this way and have no time to do it like that". And your further commit was also using the "wrong" way to call JNI, from what I conclude that you don't care.

> I wanted to know if it in general works. And for that a prototype build is needed. Because if it didn't work it's not worth to spend any time on it.

then mark as draft when creating PR. Once ready for review, mark as ready for review. Leaving a similar comment in the PR upon creation also helps.

> And the android build isn't that easy like e.g. for Linux. For non android devs the docs arent also not really clear (how to install sdk etc.). For example compiling fuzzylite doesn't work. Seen your comments in forum. But don't want mess with my system cmake install.

no idea what your problem is. Install another cmake separately or just build from source, is it not an option? I'm not an android dev FYI, just able to read docs and write C-like Java. If something doesn't work / build, open an issue. (but on CI it works, right?) I don't think I have to explain to you something like `apt install openjdk`, I'm also not a linux user btw. Ask here, ask on discord (even in PM), really...

> Btw. why do we not provide dockerfiles for building? Would be much easier with this complex building mechanism.

I don't know how to use docker and not interested in that, so feel free to do that

Laserlicht commented:
> > I was not ignoring your comments (that's why I left them open).
> 
> well, you responded with "I didn't know about this way and have no time to do it like that". And your further commit was also using the "wrong" way to call JNI, from what I conclude that you don't care.

Then sorry for created misunderstanding.

> no idea what your problem is. Install another cmake separately or just build from source, is it not an option? I'm not an android dev FYI, just able to read docs and write C-like Java. If something doesn't work / build, open an issue. (but on CI it works, right?) I don't think I have to explain to you something like `apt install openjdk`, I'm also not a linux user btw. Ask here, ask on discord (even in PM), really...

Talked about this issue: https://forum.vcmi.eu/t/fail-to-make-an-android-build/6264

> > Btw. why do we not provide dockerfiles for building? Would be much easier with this complex building mechanism.
> 
> I don't know how to use docker and not interested in that, so feel free to do that

Yeah, created a dockerfile to have a reproducible building environment here (not depending on any system library installed).

Place under: `docker/BuildAndroid.dockerfile` in vcmi dir:
```
FROM debian:trixie
WORKDIR /usr/local/app

RUN apt-get update && apt-get install -y openjdk-17-jdk python3 pipx cmake ccache ninja-build wget git sdkmanager xz-utils
RUN pipx ensurepath
RUN pipx install 'conan<2.0'

RUN ~/.local/bin/conan profile new conan --detect

RUN wget https://github.com/vcmi/vcmi-dependencies/releases/download/1.3/dependencies-android-arm64-v8a.txz
RUN tar -xf dependencies-android-arm64-v8a.txz -C ~/.conan
RUN rm dependencies-android-arm64-v8a.txz

ENV JAVA_HOME="/usr/lib/jvm/java-17-openjdk-amd64"
ENV ANDROID_HOME="/usr/lib/android-sdk"
ENV PATH="$PATH:~/.local/bin"

RUN sdkmanager --install "platform-tools"
RUN sdkmanager --install "platforms;android-34"
RUN yes | sdkmanager --licenses

RUN ~/.local/bin/conan download android-ndk/r25c@:4db1be536558d833e52e862fd84d64d75c2b3656 -r conancenter

CMD ["sh", "-c", " \
        cd /vcmi ; \
        conan install . --install-folder=conan-generated --no-imports --build=never --profile:build=default --profile:host=CI/conan/android-64-ndk ; \
        cmake --preset android-daily-release ; \
        cmake --build --preset android-daily-release \
    "]
```
build with ("vcmi" as workingdir in bash): `docker build -f docker/BuildAndroid.dockerfile -t vcmi-android-build .`
run build with (each time you want to rebuild package): `docker run -it --rm -v $PWD/:/vcmi vcmi-android-build`

Not sure if there is any interest in upstreaming this file...

kambala-decapitator commented:
btw here's how to fix fuzzylite build properly, haven't PRd it yet: https://github.com/kambala-decapitator/vcmi/commit/ca38be18d686167aea5026f4334f5ae9774f64ec

Laserlicht commented:
Addressed suggestions.

But still not able to build it correctly (missing libs in apk). But I won't pursue it any further. It's not worth looking for the problems just for this PR. If someone want to try:
```
FROM ubuntu:noble
WORKDIR /usr/local/app

RUN apt-get update && apt-get install -y openjdk-17-jdk python3 pipx cmake ccache ninja-build wget git xz-utils
RUN pipx install 'conan<2.0'
RUN pipx install 'sdkmanager==0.6.10'

RUN ~/.local/bin/conan profile new conan --detect

RUN wget https://github.com/vcmi/vcmi-dependencies/releases/download/1.3/dependencies-android-arm64-v8a.txz
RUN tar -xf dependencies-android-arm64-v8a.txz -C ~/.conan
RUN rm dependencies-android-arm64-v8a.txz

ENV JAVA_HOME="/usr/lib/jvm/java-17-openjdk-amd64"
ENV ANDROID_HOME="/usr/lib/android-sdk"
ENV GRADLE_USER_HOME="/vcmi/.cache/grandle"
ENV GENERATE_ONLY_BUILT_CONFIG=1

RUN ~/.local/bin/sdkmanager --install "platform-tools"
RUN ~/.local/bin/sdkmanager --install "platforms;android-34"
RUN yes | ~/.local/bin/sdkmanager --licenses

RUN ~/.local/bin/conan download android-ndk/r25c@:4db1be536558d833e52e862fd84d64d75c2b3656 -r conancenter

CMD ["sh", "-c", " \
    # switch to mounted dir
    cd /vcmi ; \
    # install conan stuff
    ~/.local/bin/conan install . --install-folder=conan-generated --no-imports --build=never --profile:build=default --profile:host=CI/conan/android-64-ndk ; \
    # link conan ndk that grandle can find it
    mkdir -p /usr/lib/android-sdk/ndk ; \
    ln -s -T ~/.conan/data/android-ndk/r25c/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/bin /usr/lib/android-sdk/ndk/25.2.9519653 ; \
    # build
    cmake --preset android-daily-release ; \
    cmake --build --preset android-daily-release \
"]
```

Laserlicht commented:
Fixed docker. Ready now.

IvanSavenko commented on launcher/mainwindow_moc.cpp:
```diff
@@ -264,11 +269,25 @@ void MainWindow::dropEvent(QDropEvent* event)
 void MainWindow::manualInstallFile(QString filePath)
 {
-	if(filePath.endsWith(".zip", Qt::CaseInsensitive) || filePath.endsWith(".exe", Qt::CaseInsensitive))
+#ifdef VCMI_ANDROID
+	QString realFilePath{};
+	if(filePath.contains("content://", Qt::CaseInsensitive))
+	{
+		auto path = QAndroidJniObject::fromString(filePath);
+		realFilePath = QAndroidJniObject::callStaticObjectMethod("eu/vcmi/vcmi/util/FileUtil", "getFilenameFromUri", "(Ljava/lang/String;)Ljava/lang/String;", path.object<jstring>()).toString();
+	}
+	else
+		realFilePath = filePath;
+
+#else
+	QString realFilePath = filePath;
+#endif
```
Extract to shared method? Copy-pasted at least once

Laserlicht replied:
Yes, prefered file for placing this?

IvanSavenko replied:
No string preference. 
There is apparently `helper.h/cpp` pair in Launcher. Probably in there?

IvanSavenko commented:
This should also fix failures to install game itself, not just Chronicles. Correct?

@kambala-decapitator any comments?

Laserlicht commented:
> This should also fix failures to install game itself, not just Chronicles. Correct?

I hope, yes. But cannot confirm.

kambala-decapitator commented on android/vcmi-app/src/main/java/eu/vcmi/vcmi/util/FileUtil.java:
```diff
@@ -1,18 +1,27 @@
 package eu.vcmi.vcmi.util;
 
 import android.app.Activity;
+import android.content.Context;
 import android.net.Uri;
+import android.content.ContentResolver;
+import android.provider.OpenableColumns;
+import android.database.Cursor;
 
 import androidx.annotation.Nullable;
 import androidx.documentfile.provider.DocumentFile;
 
+import org.libsdl.app.SDL;
+
 import java.io.File;
+import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.util.List;
+import java.lang.Exception;
 
+import eu.vcmi.vcmi.Const;
```
please keep imports alphabetically sorted

kambala-decapitator commented on android/vcmi-app/src/main/java/eu/vcmi/vcmi/util/FileUtil.java:
```diff
@@ -104,4 +113,50 @@ private static void copyStream(InputStream source, OutputStream target) throws I
+    {
+        try
+        {
+            String fileName = "";
```
move it outside `try` and have a single `return` in the end of the method

kambala-decapitator commented on docs/developers/Building_Android.md:
```diff
@@ -68,3 +68,17 @@ cmake --build ../build
+
+## Docker
+
+For developing it's also possible to use docker to build an android APK. As requirement only a installed docker is needed. The container image contains all other prerequisites.
```
```suggestion
For developing it's also possible to use Docker to build android APK. The only requirement is to have Docker installed. The container image contains all the other prerequisites.
```

kambala-decapitator commented on docs/developers/Building_Android.md:
```diff
@@ -68,3 +68,17 @@ cmake --build ../build
+After building the image you can compile vcmi with:
+`docker run -it --rm -v $PWD/:/vcmi vcmi-android-build`
+
+The current dockerfile is ARM64 only but can adjusted manually for ARM32.
```
please reflect it in the name of the dockerfile. Also, correct names are: armv7 for ARM32 and aarch64 for ARM64.

kambala-decapitator commented on docker/BuildAndroid.dockerfile:
```diff
@@ -0,0 +1,40 @@
+RUN sdkmanager --install "platforms;android-34"
+RUN yes | sdkmanager --licenses
+
+RUN conan download android-ndk/r25c@:4db1be536558d833e52e862fd84d64d75c2b3656 -r conancenter
```
is it not downloaded automatically when using profile with NDK?

Laserlicht replied:
Yes. But it's also here because this step is better to do in image building part, not in executing script.

kambala-decapitator commented on launcher/helper.h:
```diff
@@ -10,9 +10,11 @@
 #pragma once
 
 class QObject;
+class QString;
```
you're using QString as a fully defined type and not as a reference/pointer, hence forward declaration is incorrect - include the header instead

kambala-decapitator commented on launcher/modManager/chroniclesextractor.cpp:
```diff
@@ -226,14 +236,22 @@ void ChroniclesExtractor::installChronicles(QStringList exe)
+		logGlobal->info("Copying offline installer from '%s' to '%s'", f.toStdString(), filepath.toStdString());
+#ifdef VCMI_ANDROID
+		{
+			auto srcStr = QAndroidJniObject::fromString(f);
+			auto dstStr = QAndroidJniObject::fromString(filepath);
+			QAndroidJniObject::callStaticObjectMethod("eu/vcmi/vcmi/util/FileUtil", "copyFileFromUri", "(Ljava/lang/String;Ljava/lang/String;)V", srcStr.object<jstring>(), dstStr.object<jstring>());
```
looks copy-pasted

Laserlicht replied:
Don't see a better solution here. Suggestions?

kambala-decapitator replied:
just extract it to a separate function (e.g. put it to Helper) and call it in both places.

```cpp
void performNativeCopy(QString src, QString dst)
{
	auto srcStr = QAndroidJniObject::fromString(f);
	auto dstStr = QAndroidJniObject::fromString(filepath);
	QAndroidJniObject::callStaticObjectMethod("eu/vcmi/vcmi/util/FileUtil", "copyFileFromUri", "(Ljava/lang/String;Ljava/lang/String;)V", srcStr.object<jstring>(), dstStr.object<jstring>());
}
```

Laserlicht replied:
Ah. Misunderstood you first.

Makes sense!

kambala-decapitator commented on android/vcmi-app/src/main/java/eu/vcmi/vcmi/util/FileUtil.java:
```diff
@@ -104,4 +113,50 @@ private static void copyStream(InputStream source, OutputStream target) throws I
+    {
+        try
+        {
+            final Context ctx = SDL.getContext();
```
pass context from C++ as main activity, Qt provides a method to get it. Treat SDL as non-existant in the launcher.

kambala-decapitator commented on android/vcmi-app/src/main/java/eu/vcmi/vcmi/util/FileUtil.java:
```diff
@@ -104,4 +113,50 @@ private static void copyStream(InputStream source, OutputStream target) throws I
+        {
+            String fileName = "";
+
+            final Context ctx = SDL.getContext();
```
same

kambala-decapitator commented:
btw are there any checks that the passed exe is actually chronicles and not something random that can be extracted?

Laserlicht commented:
> btw are there any checks that the passed exe is actually chronicles and not something random that can be extracted?

Yes, there is a logic after extracting (there was also one before, but @IvanSavenko removed it to simplify process).

Laserlicht commented:
Done open points. Tested again on my device - works.

kambala-decapitator commented on launcher/helper.cpp:
```diff
@@ -44,4 +49,30 @@ void enableScrollBySwiping(QObject * scrollTarget)
+	auto dstStr = QAndroidJniObject::fromString(dst);
+	QAndroidJniObject::callStaticObjectMethod("eu/vcmi/vcmi/util/FileUtil", "copyFileFromUri", "(Ljava/lang/String;Ljava/lang/String;Landroid/content/Context;)V", srcStr.object<jstring>(), dstStr.object<jstring>(), QtAndroid::androidContext().object());
+#else
+	QFile(src).copy(dst);
```
```suggestion
	QFile::copy(src, dst);
```

kambala-decapitator reviewed (approved):
all good now, just do a minor adjustment please
